cpp/ql/src/semmle/code/cpp/controlflow/StackVariableReachability.qll

```diff
@@ -80,46 +268,198 @@ abstract class StackVariableReachability extends string {
         j > i and
         sink = bb.getNode(j) and
         isSink(sink, v) and
-        not exists(int k | isBarrier(bb.getNode(k), v) | k in [i + 1 .. j - 1])
+        not isBarrier(bb.getNode(pragma[only_bind_into]([i + 1 .. j - 1])), v)
       )
       or
       not exists(int k | isBarrier(bb.getNode(k), v) | k > i) and
-      bbSuccessorEntryReaches(bb, v, sink, _)
+      bbSuccessorEntryReaches(source, bb, v, sink, _)
+    )
+  }
+
+  private Condition getASinkCondition(SemanticStackVariable v) {
+    exists(BasicBlock bb |
+      revBbEntryReachesLocally(bb, v, _, this) and
+      result
+          .(LogicalCondition)
+          .getCondition()
+          .controls(bb, result.(LogicalCondition).getTruthValue())
+    )
+  }
+
+  private Condition getABarrierCondition(SemanticStackVariable v) {
+    exists(BasicBlock bb |
+      isBarrier(bb.getANode(), v) and
+      result
+          .(LogicalCondition)
+          .getCondition()
+          .controls(bb, result.(LogicalCondition).getTruthValue())
+    )
+  }
+
+  private predicate successorExitsLoop(BasicBlock pred, BasicBlock succ, Loop loop) {
+    pred.getASuccessor() = succ and
+    bbDominates(loop.getStmt(), pred) and
+    not bbDominates(loop.getStmt(), succ)
+  }
+
+  private predicate successorExitsFirstDisjunct(BasicBlock pred, BasicBlock succ) {
+    exists(LogicalOrExpr orExpr | orExpr instanceof GuardCondition |
+      pred.getAFalseSuccessor() = succ and
+      pred.contains(orExpr.getLeftOperand())
+    )
+  }
+
+  /** Holds if `cond` is a condition with a known truth value in `bb`. */
```

Comment:
```suggestion
  /** Gets a condition with a known truth value in `bb`. */
```
What are `source` and `v`?

Reply:
I've added a comment explaining the role of `source` and `v` in 656ff4aee906cee0a472fafd2d5567d9425fb29c.